### ISSUE TITLE
question on the `block_table` and `cache_seqlens` in `flash_attn_with_kvcache`

### ISSUE THREAD

**Issue by chakpongchung** [Dao-AILab/flash-attention#1314]
https://github.com/Dao-AILab/flash-attention/blob/478ee666cccbd1b8f63648633003059a8dc6827d/tests/test_flash_attn.py#L2066

Could you elaborate more on the block table argument here? I am trying to find an example to show how `flash_attn_with_kvcache` should be used. Specifically, why user needs to care about the logical block/physical block mapping when passing the KV cache since it is updated in place by this function? How should we construct the block table based on the KV shape? I assume the block table shape is dependent on the KV shape only.

**Comment by tridao** [comment]
You can read the function docstring
https://github.com/Dao-AILab/flash-attention/blob/478ee666cccbd1b8f63648633003059a8dc6827d/flash_attn/flash_attn_interface.py#L1492

Typically this is used with a cache manager (e.g. in vLLM) that manages when to allocate & free the blocks and construct the block table. Such cache manager is not implemented here since it depends on how you build the inference engine.

**Comment by chakpongchung** [comment]
Where can we find a better example to show us how to use this API with block table?
For example, in vLLM (0.6.1.post2) what does it mean by having a block_table with value
```
tensor([[127],
        [126],
        [125]], device='cuda:0', dtype=torch.int32)
```
?

In this case, the prompt for the model gpt2 is 

```
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",]
```
and 
```
sampling_params = SamplingParams(temperature=0, top_p=1, top_k=1,max_tokens=1024)

```

**Comment by tridao** [comment]
You can read the test code (https://github.com/Dao-AILab/flash-attention/blob/641db759ab7168e472909bc9ff1eda4a329de34f/tests/test_flash_attn.py#L1907) to see how block table is used. Setting breakpoint and printing out values could be helpful

**Comment by chakpongchung** [comment]
Thanks. I found the answer.

**Comment by chakpongchung** [comment]
A follow-up question :

I need to know more about the seq_len usage inside the function you previously pointed out
https://github.com/Dao-AILab/flash-attention/blob/641db759ab7168e472909bc9ff1eda4a329de34f/flash_attn/flash_attn_interface.py#L1547

which will in turn execute this cuda kernel:
https://github.com/Dao-AILab/flash-attention/blob/641db759ab7168e472909bc9ff1eda4a329de34f/csrc/flash_attn/flash_api.cpp#L1241

(confused about this comment which says batch_size in the comment about seq_len)
To your knowledge, is it possible to use `cuda-gdb` to debug this test? If not, how do you normally set break-point and print in it? (don't want to add print and then recompile)
```
cuda-gdb --args pytest -v tests/test_flash_attn.py::test_flash_attn_kvcache
```

**Comment by tridao** [comment]
The tests should fully describe the usage. 
I always add print and recompile

**Comment by tridao** [comment]
I.e the tests compare the output of the cuda kernel w a reference impl. You just need to trace through the ref impl (in PyTorch) to know how block table is used.

**Comment by chakpongchung** [comment]
I see that it is comparing the result from flashAttn with the one from the ref attention impl using pytorch (e.g. softmax followed by the `torch.einsum`).


Why tracing through this ref impl helps understanding `block_table` and `cache_seqlens`? They are not used there. 

I updated the issue title to point out the confusion I have when using the `flash_attn_with_kvcache` in case there is any confusion caused by my earlier question. My intent is to construct the appropriate k cache and v cache so they can be used by  `flash_attn_with_kvcache`

**Comment by aleksmark98** [comment]
@chakpongchung I suggest you take a look at a vllm test of flash attention. They are comparing it against their pytorch implementation [`ref_paged_attn`](https://github.com/vllm-project/vllm/blob/b590adfdc15fc716f6d120aeefeb587f491f8fce/tests/kernels/test_flash_attn.py#L24), where you can read through how `block_tables` and `cache_seqlens` are used to index into key and value caches.
